File: kotlin-node/karakum/plugins/convertUtilityTypes.js
import ts from "typescript"

export default function (node, context, render) {
    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && (
            node.typeName.text === "Pick"
            || node.typeName.text === "Omit"
        )
    ) {
        return render(node.typeArguments[0])
    }
    return null
}

File: kotlin-typescript/karakum/src/jsMain/kotlin/typescript/karakum/plugins/convertUtilityTypes.kt
package typescript.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.asArray
import typescript.isIdentifier
import typescript.isTypeReferenceNode

val convertUtilityTypes = createPlugin { node, _, render ->
    nullable {
        ensure(isTypeReferenceNode(node))

        val typeName = node.typeName
        ensure(isIdentifier(typeName))
        ensure(typeName.text == "NonNullable")

        val firstTypeArgument = ensureNotNull(node.typeArguments?.asArray()?.firstOrNull())

        "${render(firstTypeArgument)} & Any"
    }
}
